lib/Checks/TrailingWhitespaceCheck.vala

```diff
@@ -28,17 +28,17 @@ public class ValaLint.Checks.TrailingWhitespaceCheck : Check {
         return _("Checks for whitespaces at the end of lines");
     }
 
-    public override bool check_line (Gee.ArrayList<FormatMistake?> mistake_list, int line_index, string line) {
-        string clean_line = line.chomp ();
-
-        if (clean_line != line) {
-            string mistake = _("Unexpected whitespace at end of line");
-
-            mistake_list.add ({ this, line_index, clean_line.length, mistake });
-
-            return true;
+    public override void check (Gee.ArrayList<ParseResult? > parse_result, Gee.ArrayList<FormatMistake? > mistake_list) {
+        foreach (ParseResult r in parse_result) {
+            if (r.type == ParseType.Default) {
+                add_regex_mistake (this, """ \n""", "Unexpected whitespace at end of line", r, mistake_list);
+            }
         }
 
-        return false;
+        // Check for whitespace at last line
+        ParseResult r_last = parse_result.last ();
+        if (r_last.type == ParseType.Default) {
+            add_regex_mistake (this, """ $""", "Unexpected whitespace at end of last line", r_last, mistake_list);
```

Comment:
Same here with the horizontal whitespace

Reply:
As discussed above, with a multiline regex, you'd only need one check here, because you could check against the end of the line with the `$` character instead of having the `\n` check too.

Reply:
The problem is that because the parser splits the file content, the end of a string does not need to be the end of a line. For example the code `string a = "lorem";` is split into three parts
1. `string a = `
2. `"lorem"`
3. `;`

If we use only the `$` character for end of line/end of string with the MultiLine option, then the last whitespace in 1. is matched as well.